### PR TITLE
Geant4 sensitive detectors update 3

### PR DESCRIPTION
**Pull request by civanch**
This PR is made instead of #21767 and #21797 
Geant4 calorimeter sensitive detectors are updated:
1) use const G4Step* pointer in protected and private interfaces where possible
2) Two main methods for sub-detectors : getFromLibrary(..) amd getEnergyDeposit(.. )
3) use std::unique_ptr, auto, and other C++11 features in several places
4) reduce number of protected class members
5) reduced number of type conversions
6) optimized ECalSD run time methods - perform one search over map of logical volumes instead of 3,
use const pointers and references
7) remove usage of G4ParticleTable at initialisation use instead static utility
8) removed duplicate codes 
9) optimised LogInfo and LogDebug - verbose output is substantially smaller but keep similar information
10) removed old commented lines
11) applied corrections suggested by @kpedro88 and @VinInn 

Correctness of this PR means identical results for all WFs.

### CODE REVIEW

**Comment by civanch** [comment]
please test

**Comment by civanch** [comment]
please test

**Comment by civanch** [comment]
please test

File: SimG4CMS/Forward/src/CastorSD.cc [hunk 1]
```diff
@@ -476,20 +463,22 @@ bool CastorSD::getFromLibrary(const G4Step* aStep) {
...
-    
     if (currentLV == lvC3EF || currentLV == lvC4EF || currentLV == lvC3HF ||
         currentLV == lvC4HF) {
       G4double edep     = aStep->GetTotalEnergyDeposit();
       G4double beta     = preStepPoint->GetBeta();
       G4double charge   = preStepPoint->GetCharge();
       double bThreshold = 0.67;
-      if(edep == 0.0 && charge != 0.0 && beta > bThreshold) { G4Poisson(0.0); }
+      if(edep == 0.0 && charge != 0.0 && beta > bThreshold) { 
+	G4Poisson(0.0);
```
Comment (kpedro88) [inline SimG4CMS/Forward/src/CastorSD.cc]: What is the point of calling `G4Poisson` here if the return value is ignored?

**Comment by kpedro88** [comment]
@civanch still seeing a lot of differences/fluctuations in the comparisons

**Comment by civanch** [comment]
@kpedro88 , unfortunately, yes, it is improving but I have to find all places were regression is broken to be sure that nothing bad is coming with the redesign.

**Comment by civanch** [comment]
please test

**Comment by civanch** [comment]
please test

**Comment by civanch** [comment]
please test

**Comment by civanch** [comment]
please test

**Comment by civanch** [comment]
As discussed at SIM meeting February,6 we should change the order of modifications. This PR should be closed.

**Pull request closed without merging**